sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -325,200 +329,202 @@ export class BatchingReceiverLite {
+      // Drain any pending credits.
+      if (receiver.isOpen() && receiver.credit > 0) {
+        logger.verbose(`${loggingPrefix} Draining leftover credits(${receiver.credit}).`);
+
+        // setting .drain and combining it with .addCredit results in (eventually) sending
+        // a drain request to Service Bus. When the drain completes rhea will call `onReceiveDrain`
+        // at which point we'll wrap everything up and resolve the promise.
+        receiver.drain = true;
+        receiver.addCredit(1);
+      } else {
         logger.verbose(
-          `${loggingPrefix} Drained, resolving receiveMessages() with ${brokeredMessages.length} messages.`
+          `${loggingPrefix} Resolving receiveMessages() with ${brokeredMessages.length} messages.`
         );
+        resolveImmediately(brokeredMessages);
+      }
+    };
 
-        // NOTE: through rhea-promise most of our event handlers are made asynchronous by calling setTimeout(emit).
-        // However, a small set (*error and drain) execute immediately. This can lead to a situation where the logical
-        // ordering of events is correct but the execution order is incorrect because the events are not all getting
-        // put into the task queue the same way.
-        // So this call, while odd, just ensures that we resolve _after_ any already-queued onMessage handlers that may
-        // be waiting in the task queue.
-        setTimeout(() => {
-          cleanupBeforeResolveOrReject();
-          resolve(brokeredMessages);
-        });
-      };
-
-      cleanupBeforeResolveOrReject = (): void => {
-        if (receiver != null) {
-          receiver.removeListener(ReceiverEvents.receiverError, onError);
-          receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
-          receiver.session.removeListener(SessionEvents.sessionError, onError);
-          receiver.removeListener(ReceiverEvents.receiverClose, onClose);
-          receiver.session.removeListener(SessionEvents.sessionClose, onClose);
-          receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
+    // Action to be performed on the "message" event.
+    const onReceiveMessage: OnAmqpEventAsPromise = async (context: EventContext) => {
+      // TODO: this appears to be aggravating a bug that we need to look into more deeply.
+      // The same timeout+drain sequence should work fine for receiveAndDelete but it appears
+      // to cause problems.
+      if (this._receiveMode === "peekLock") {
+        if (brokeredMessages.length === 0) {
+          // We'll now remove the old timer (which was the overall `maxWaitTimeMs` timer)
+          // and replace it with another timer that is a (probably) much shorter interval.
+          //
+          // This allows the user to get access to received messages earlier and also gives us
+          // a chance to have fewer messages internally that could get lost if the user's
+          // app crashes in receiveAndDelete mode.
```

Comment:
Cleanup: This part of the comment can be removed since this logic is not ran in receiveAndDelete mode.

Reply:
Done.